README
NAME

    Mock::Sub - Mock package, object and standard subroutines, with unit
    testing in mind.

SYNOPSIS

        # see EXAMPLES for a full use case and caveats
    
        use Mock::Sub;
    
        # create the parent mock object
    
        my $mock = Mock::Sub->new;
    
        # mock some subs...
    
        my $foo = $mock->mock('Package::foo');
        my $bar = $mock->mock('Package::bar');
    
        # wait until a mocked sub is called
    
        Package::foo();
    
        # then...
    
        $foo->name;         # name of sub that's mocked
        $foo->called;       # was the sub called?
        $foo->called_count; # how many times was it called?
        $foo->called_with;  # array of params sent to sub
    
        # have the mocked sub return something when it's called (list or scalar).
    
        $foo->return_value(1, 2, {a => 1});
        my @return = Package::foo;
    
        # have the mocked sub perform an action
    
        $foo->side_effect( sub { die "eval catch" if @_; } );
    
        eval { Package::foo(1); };
        like ($@, qr/eval catch/, "side_effect worked with params");
    
        # extract the parameters the sub was called with
    
        my @args = $foo->called_with;
    
        # reset the mock object for re-use within the same scope
    
        $foo->reset;
    
        # restore original functionality to the sub
    
        $foo->unmock;
    
        # re-mock a previously unmock()ed sub
    
        $foo->remock;
    
        # check if a sub is mocked
    
        my $state = $foo->mocked_state;
    
        # mock out a CORE:: function. Be warned that this *must* be done within
        # compile stage (BEGIN), and the function can NOT be unmocked prior
        # to the completion of program execution
    
        my ($mock, $caller);
    
        BEGIN {
            $mock = Mock::Sub->new;
            $caller = $mock->mock('caller');
        };
    
        $caller->return_value(55);
        caller(); # mocked caller() called

DESCRIPTION

    Easy to use and very lightweight module for mocking out sub calls. Very
    useful for testing areas of your own modules where getting coverage may
    be difficult due to nothing to test against, and/or to reduce test run
    time by eliminating the need to call subs that you really don't want or
    need to test.

EXAMPLE

    Here's a full example to get further coverage where it's difficult if
    not impossible to test certain areas of your code (eg: you have if/else
    statements, but they don't do anything but call other subs. You don't
    want to test the subs that are called, nor do you want to add
    statements to your code).

    Note that if the end subroutine you're testing is NOT Object Oriented
    (and you're importing them into your module that you're testing), you
    have to mock them as part of your own namespace (ie. instead of
    Other::first, you'd mock MyModule::first).

       # module you're testing:
    
        package MyPackage;
        
        use Other;
        use Exporter qw(import);
        @EXPORT_OK = qw(test);
       
        my $other = Other->new;
    
        sub test {
            my $arg = shift;
            
            if ($arg == 1){
                # how do you test this?... there's no return etc.
                $other->first();        
            }
            if ($arg == 2){
                $other->second();
            }
        }
    
        # your test file
    
        use MyPackage qw(test);
        use Mock::Sub;
        use Test::More tests => 2;
    
        my $mock = Mock::Sub->new;
    
        my $first = $mock->mock('Other::first');
        my $second = $mock->mock('Other::second');
    
        # coverage for first if() in MyPackage::test
        test(1);
        is ($first->called, 1, "1st if() statement covered");
    
        # coverage for second if()
        test(2);
        is ($second->called, 1, "2nd if() statement covered");

MOCK OBJECT METHODS

 new(%opts)

    Instantiates and returns a new Mock::Sub object, ready to be used to
    start creating mocked sub objects.

    Optional options:

    return_value => $scalar

      Set this to have all mocked subs created with this mock object return
      anything you wish (accepts a single scalar only. See return_value()
      method to return a list and for further information). You can also
      set it in individual mocks only (see return_value() method).

    side_effect => $cref

      Set this in new() to have the side effect passed into all child mocks
      created with this object. See side_effect() method.

 mock('sub', %opts)

    Instantiates and returns a new mock object on each call. 'sub' is the
    name of the subroutine to mock (requires full package name if the sub
    isn't in main::).

    The mocked sub will return undef if a return value isn't set, or a side
    effect doesn't return anything.

    Optional parameters:

    See new() for a description of the parameters. Both the return_value
    and side_effect parameters can be set in this method to individualize
    each mock object, and will override the global configuration if set in
    new().

    There's also return_value() and side_effect() methods if you want to
    set, change or remove these values after instantiation of a child sub
    object.

 mocked_subs

    Returns a list of all the names of the subs that are currently mocked
    under the parent mock object.

 mocked_objects

    Returns a list of all sub objects underneath the parent mock object,
    regardless if its sub is currently mocked or not.

 mocked_state('Sub::Name')

    Returns 1 if the sub currently under the parent mock object is mocked
    or not, and 0 if not. Croaks if there hasn't been a child sub object
    created with this sub name.

SUB OBJECT METHODS

    These methods are for the children mocked sub objects returned from the
    parent mock object. See "MOCK OBJECT METHODS" for methods related to
    the parent mock object.

 unmock

    Restores the original functionality back to the sub, and runs reset()
    on the object.

 remock

    Re-mocks the sub within the object after calling unmock on it (accepts
    the side_effect and return_value parameters).

 called

    Returns true (1) if the sub being mocked has been called, and false (0)
    if not.

 called_count

    Returns the number of times the mocked sub has been called.

 called_with

    Returns an array of the parameters sent to the subroutine. confess()s
    if we're called before the mocked sub has been called.

 mocked_state

    Returns true (1) if the sub the object refers to is currently mocked,
    and false (0) if not.

 name

    Returns the name of the sub being mocked.

 side_effect($cref)

    Add (or change/delete) a side effect after instantiation.

    Send in a code reference containing an action you'd like the mocked sub
    to perform.

    The side effect function will receive all parameters sent into the
    mocked sub.

    You can use both side_effect() and return_value() params at the same
    time. side_effect will be run first, and then return_value. Note that
    if side_effect's last expression evaluates to any value whatsoever
    (even false), it will return that and return_value will be skipped.

    To work around this and have the side_effect run but still get the
    return_value thereafter, write your cref to evaluate undef as the last
    thing it does: sub { ...; undef; }.

 return_value

    Add (or change/delete) the mocked sub's return value after
    instantiation. Can be a scalar or list. Send in undef to remove
    previously set values.

 reset

    Resets the functional parameters (return_value, side_effect), along
    with called() and called_count() back to undef/false. Does not restore
    the sub back to its original state.

NOTES

    This module has a backwards parent-child relationship. To use, you
    create a mock object using "PARENT MOCK OBJECT METHODS" new and mock
    methods, thereafter, you use the returned mocked sub object METHODS to
    perform the work.

    The parent mock object retains certain information and statistics of
    the child mocked objects (and the subs themselves).

    To mock CORE::GLOBAL functions, you *must* initiate within a BEGIN
    block (see SYNOPSIS for details). It is important that if you mock a
    CORE sub, it can't and won't be returned to its original state until
    after the entire program process tree exists. Period.

    I didn't make this a Test:: module (although it started that way)
    because I can see more uses than placing it into that category.

AUTHOR

    Steve Bertrand, <steveb at cpan.org>

BUGS

    Please report any bugs or requests at
    https://github.com/stevieb9/mock-sub/issues

REPOSITORY

    https://github.com/stevieb9/mock-sub

BUILD RESULTS

    CPAN Testers: http://matrix.cpantesters.org/?dist=Mock-Sub

SUPPORT

    You can find documentation for this module with the perldoc command.

        perldoc Mock::Sub

ACKNOWLEDGEMENTS

    Python's MagicMock module.

LICENSE AND COPYRIGHT

    Copyright 2016 Steve Bertrand.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Mock-Sub

1.10 UNREL
        - add semi-colon to example in SYNOPSIS (fixes #23)

1.09    2017-12-28
        - setting a return_value or side_effect when mocking one sub doesn't
          affect any other subs mocked (fixes #25)
        - params passed into a mocked sub will only be returned if
          return_value is set to "params" (closes #26)

1.08    2017-12-21
        - the mocked subs now return the parameters that were sent in to it if
          return_value or side_effect have not been set (closes #24)

1.07    2016-10-05
        - POD fix (closes #20)
        - changed croak() to confess()
        - you can now add "no_warnings => 1" to the 'use Mock::Sub' line to
          disable warnings about mocking non-existent subs (closes #22)

1.06    2016-01-23
        - new remock() method does what Child::mock() used to do.
        - Child::mock() renamed to Child::_mock(), and is now private
          (fixes #17)
        - more documentation cleanup

1.05    2016-01-22
        - added tests for mocked_objects()
        - Child.pm now has basic pod (mostly to satisfy pod tests) (closes #15)
        - added .list to MANIFEST.SKIP to ignore RHEL vendor specific build
          files (reported by Denis Fateyev) (also #15)
        - this was previously fixed (closes #13)

1.04    2016-01-08
        - MANIFEST updates

1.03    2015-12-10
        - completely reworked distribution into two modules (parent and Child)
        - we now have the ability to mock out CORE::GLOBAL functions
          (eg: caller()). You MUST do the mock within a BEGIN block

1.02    2015-12-09
        - fixed bug where if the mocked sub isn't called, the original sub
          wasn't being restored
        - called_count() now returns 0 instead of undef if there isn't a count
        - fix issue #11: called_count wasn't being reset to 0 after reset()
        - side_effect param can now be sent into new() which will be retained
          in all child objects created by mock()
        - return_value param can be send into new(), passed into all child
          objects
        - added mocked_objects(), returns a list of objects that have mocked
          subs (if unmocked(), obj won't be returned)
        - added mocked_names(), same as mocked_objects() but returns the sub
          names only
        - mocked_state(), returns bool whether the named sub is currently
          mocked

1.01    2015-12-08
        - DESTROY() always calls unmock() now. The option to disable this has
          been removed

0.11    2015-12-08
        - if mock() is called in void context, we'll croak (except in the
          instance where you're re-mocking an unmock()ed sub with the same
          object)
        - fixes bug where DESTROY() wasn't able to properly clean up prior
          to END

0.10    2015-12-07
        - all bugs fixed in this release reported by Joel Berger (jberger)
        - fix issue #5: we now pass in any parameters passed to the mocked
          function into the side_effect cref
        - fix issue #6: return_value() now accepts lists (however, using the
          param version of return_value, only a scalar is accepted).
        - fix issue #8: called_with() now die()s if it's called before the
          mocked sub has been
        - fix issue #7: mocked sub (and the object) is now much more context
          aware
        - fix issue #9: DESTROY() now unmock()s by default
          (keep_mock_on_destroy param to override)

0.09    2015-12-01
        - mocking a non-existent sub is now allowed, and spits out a warning
          message
        - unmock() removes the symtab entry if mock() created a fake
          (non-existent) sub entry in the symbol table
        - added side_effect() and return_value()

0.08    2015-11-30
        - fixes bug #3: add eval() in side_effect in SYNOPSIS
        - fixes bug #2: having side_effect after return_value breaks t/06
        - fixes bug #1: add called_with(), to extract the args passed in to
          the mocked sub
        - fixes bug #4: completely validate that the sub param is a valid sub
        - 'main::' is now prepended to subs if they don't have a package
        - side_effect now properly returns last expression evaluated
        - side_effect and return_value can now be sent in simultaneously

0.07    2015-11-29
        - code cleanup
        - pod updates 
        - opened bug 1: need to add with_args()

0.04    2015-11-29
        - added name()
        - many more tests
        - major POD updates/fixes

0.02    2015-11-29
        - renamed project from Test::MockSub to Mock::Sub
        - added numerous test files
        - added initial POD

0.01    2015-11-29
        - initial import



t/data/A.pm
package A;

sub new {
    return bless {}, shift;
}

sub foo {
    print "in A::foo\n";
}
sub bar {
    print "in A::bar\n";
}
sub baz {
    print "in A::baz\n";
}
1;


t/data/B.pm
package B;

use lib '.';
use A;

sub test {
    my $obj = A->new;
    $obj->foo;
}
sub test2 {
    my $obj = A->new;
    $obj->bar;
}
sub test3 {
    my $obj = A->new;
    $obj->baz;
}

1;


t/data/One.pm
package One;

sub new {
    return bless {}, shift;
}

sub foo {
    print "in One::foo\n";
}
sub bar {
    print "in One::bar\n";
}
sub baz {
    print "in One::baz\n";
}
1;


Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'Mock::Sub',
    AUTHOR           => q{Steve Bertrand <steveb@cpan.org>},
    VERSION_FROM     => 'lib/Mock/Sub.pm',
    ABSTRACT_FROM    => 'lib/Mock/Sub.pm',
    LICENSE          => 'perl_5',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.006,
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources   => {
            bugtracker => 'https://github.com/stevieb9/mock-sub/issues',
            repository => {
                type => 'git',
                url => 'https://github.com/stevieb9/mock-sub.git',
                web => 'https://github.com/stevieb9/mock-sub',
            },
        },
    },
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
        'Carp' => 0,
        'Scalar::Util' => 0,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Mock-Sub-*' },
);


lib/Mock/Sub/Child.pm
package Mock::Sub::Child;
use 5.006;
use strict;
use warnings;

use Carp qw(confess);
use Scalar::Util qw(weaken);

our $VERSION = '1.10';

sub new {
    my $self = bless {}, shift;
    %{ $self } = @_;

    if ($self->{side_effect}){
        $self->_check_side_effect($self->{side_effect});
    }
    return $self;
}
sub _mock {
    my $self = shift;

    # throw away the sub name if it's sent in and we're not called
    # by Mock::Sub::mock()

    my $sub_passed_in;
    if ($_[0] && $_[0] =~ /::/){
        $sub_passed_in = 1;
    }

    my $caller = (caller(1))[3] || '';
    
    if ($caller ne 'Mock::Sub::mock' && $sub_passed_in){
        undef @_;
        if(ref($self) eq 'Mock::Sub::Child' && ! $self->{name}){
            confess "can't call mock() on a child object before it is already " .
                  "initialized with the parent mock object. ";
        }
    }

    if ($caller ne 'Mock::Sub::mock' && $caller ne 'Mock::Sub::Child::remock'){
        confess "the _mock() method is not a public API call. For re-mocking " .
              "an existing sub in an existing sub object, use remock().\n";
    }

    my $sub = $self->name || shift;

    my %p = @_;
    for (keys %p){
        $self->{$_} = $p{$_};
    }

    if ($sub !~ /::/) {
        my $core_sub = "CORE::" . $sub;

        if (defined &$core_sub && ${^GLOBAL_PHASE} eq 'START') {
            warn "WARNING! we're attempting to override a global core " .
                 "function. You will NOT be able to restore functionality " .
                 "to this function.";

            $sub = "CORE::GLOBAL::" . $sub;
        }
        else {
            $sub = "main::$sub" if $sub !~ /::/;
        }
    }

    my $fake;

    if (! exists &$sub && $sub !~ /CORE::GLOBAL/) {
        $fake = 1;
        if (! $self->_no_warn) {
            warn "\n\nWARNING!: we've mocked a non-existent subroutine. ".
                    "the specified sub does not exist.\n\n";
        }
    }

    $self->_check_side_effect($self->{side_effect});

    if (defined $self->{return_value}){
        push @{ $self->{return} }, $self->{return_value};
    }

    $self->{name} = $sub;
    $self->{orig} = \&$sub if ! $fake;

    $self->{called_count} = 0;

    {
        no strict 'refs';
        no warnings 'redefine';

        my $mock = $self;
        weaken $mock;

        *$sub = sub {

            @{ $mock->{called_with} } = @_;
            ++$mock->{called_count};

            if ($mock->{side_effect}) {
                if (wantarray){
                    my @effect = $mock->{side_effect}->(@_);
                    return @effect if @effect;
                }
                else {
                    my $effect = $mock->{side_effect}->(@_);
                    return $effect if defined $effect;
                }
            }

            return if ! defined $mock->{return};

            if ($mock->{return}[0] && $mock->{return}[0] eq 'params'){
                return ! wantarray ? $_[0] : @_;
            }
            else {
                return ! wantarray && @{ $mock->{return} } == 1
                    ? $mock->{return}[0]
                    : @{ $mock->{return} };
            }
        };
    }
    $self->{state} = 1;

    return $self;
}
sub remock {
    shift->_mock(@_);
}
sub unmock {
    my $self = shift;
    my $sub = $self->{name};

    {
        no strict 'refs';
        no warnings 'redefine';

        if (defined $self->{orig} && $sub !~ /CORE::GLOBAL/) {
            *$sub = \&{ $self->{orig} };
        }
        else {
            undef *$sub if $self->{name};
        }
    }

    $self->{state} = 0;
    $self->reset;
}
sub called {
    return shift->called_count ? 1 : 0;
}
sub called_count {
    return shift->{called_count} || 0;
}
sub called_with {
    my $self = shift;
    if (! $self->called){
        confess "\n\ncan't call called_with() before the mocked sub has " .
            "been called. ";
    }
    return @{ $self->{called_with} };
}
sub name {
    return shift->{name};  
}
sub reset {
    for (qw(side_effect return_value return called called_count called_with)){
        delete $_[0]->{$_};
    }
}
sub return_value {
    my $self = shift;
    @{ $self->{return} } = @_;
}
sub side_effect {
    $_[0]->_check_side_effect($_[1]);
    $_[0]->{side_effect} = $_[1];
}
sub _check_side_effect {
    if (defined $_[1] && ref $_[1] ne 'CODE') {
        confess "\n\nside_effect parameter must be a code reference. ";
    }
}
sub mocked_state {
    return shift->{state};
}
sub _no_warn {
    return $_[0]->{no_warnings};
}
sub DESTROY {
    $_[0]->unmock;
}
sub _end {}; # vim fold placeholder

__END__

=head1 NAME

Mock::Sub::Child - Provides for Mock::Sub

=head1 METHODS

Please refer to the C<Mock::Sub> parent module for full documentation. The
descriptions here are just a briefing.

=head2 new

This method can only be called by the parent C<Mock::Sub> module.

=head2 called

Returns bool whether the mocked sub has been called yet.

=head2 called_count

Returns an integer representing the number of times the mocked sub has been
called.

=head2 called_with

Returns a list of arguments the mocked sub was called with.

=head2 mock

This method should only be called by the parent mock object. You shouldn't be
calling this.

=head2 remock

Re-mocks an unmocked sub back to the same subroutine it was originally mocked
with.

=head2 mocked_state

Returns bool whether the sub the object represents is currently mocked or not.

=head2 name

Returns the name of the sub this object is mocking.

=head2 return_value

Send in any values (list or scalar) that you want the mocked sub to return when
called.

=head2 side_effect

Send in a code reference with any actions you want the mocked sub to perform
after it's been called.

=head2 reset

Resets all state of the object back to default (does not unmock the sub).

=head2 unmock

Restores original functionality of the mocked sub, and calls C<reset()> on the
object.

=cut
1;



lib/Mock/Sub.pm
package Mock::Sub;
use 5.006;
use strict;
use warnings;

use Carp qw(confess);
use Mock::Sub::Child;
use Scalar::Util qw(weaken);

our $VERSION = '1.10';

my %opts;

sub import {
    my ($class, %args) = @_;
    %opts = %args;
}
sub new {
    my $self = bless {}, shift;
    %{ $self } = @_;

    for (keys %opts){
        $self->{$_} = $opts{$_};
    }
    return $self;
}
sub mock {
    my ($self, $sub, %p) = @_;

    if (ref($self) ne 'Mock::Sub'){
        confess
            "calling mock() on the Mock::Sub class is no longer permitted. " .
            "create a new mock object with Mock::Sub->new;, then call mock " .
            "with my \$sub_object = \$mock->mock('sub_name'); ";
    }

    if (! defined wantarray){
        confess "\n\ncalling mock() in void context isn't allowed. ";
    }

    my $child = Mock::Sub::Child->new(no_warnings => $self->{no_warnings});

    my $side_effect = defined $p{side_effect}
        ? $p{side_effect}
        : $self->{side_effect};

    my $return_value = defined $p{return_value}
        ? $p{return_value}
        : $self->{return_value};

    $child->side_effect($side_effect);
    $child->return_value($return_value);

    $self->{objects}{$sub}{obj} = $child;
    $child->_mock($sub);

    # remove the REFCNT to the child, or else DESTROY won't be called
    weaken $self->{objects}{$sub}{obj};

    return $child;
}
sub mocked_subs {
    my $self = shift;

    my @names;

    for (keys %{ $self->{objects} }) {
        if ($self->mocked_state($_)){
            push @names, $_;
        }
    }
    return @names;
}
sub mocked_objects {
    my $self = shift;

    my @mocked;
    for (keys %{ $self->{objects} }){
        push @mocked, $self->{objects}{$_}{obj};
    }
    return @mocked;
}
sub mocked_state {
    my ($self, $sub) = @_;

    if (! $sub){
        confess "calling mocked_state() on a Mock::Sub object requires a sub " .
              "name to be passed in as its only parameter. ";
    }

    eval {
        my $test = $self->{objects}{$sub}{obj}->mocked_state();
    };
    if ($@){
        confess "can't call mocked_state() on the class if the sub hasn't yet " .
              "been mocked. ";
    }
    return $self->{objects}{$sub}{obj}->mocked_state;
}
sub DESTROY {
}
sub __end {}; # vim fold placeholder

1;
=head1 NAME

Mock::Sub - Mock package, object and standard subroutines, with unit testing in mind.

=for html
<a href="http://travis-ci.org/stevieb9/mock-sub"><img src="https://secure.travis-ci.org/stevieb9/mock-sub.png"/>
<a href='https://coveralls.io/github/stevieb9/mock-sub?branch=master'><img src='https://coveralls.io/repos/stevieb9/mock-sub/badge.svg?branch=master&service=github' alt='Coverage Status' /></a>

=head1 SYNOPSIS

    # see EXAMPLES for a full use case and caveats

    use Mock::Sub;

    # disable warnings about mocking non-existent subs

    use Mock::Sub no_warnings => 1;

    # create the parent mock object

    my $mock = Mock::Sub->new;

    # mock some subs...

    my $foo = $mock->mock('Package::foo');
    my $bar = $mock->mock('Package::bar');

    # wait until a mocked sub is called

    Package::foo();

    # then...

    $foo->name;         # name of sub that's mocked
    $foo->called;       # was the sub called?
    $foo->called_count; # how many times was it called?
    $foo->called_with;  # array of params sent to sub

    # have the mocked sub return something when it's called (list or scalar).

    $foo->return_value(1, 2, {a => 1});
    my @return = Package::foo;

    # have the mocked sub perform an action

    $foo->side_effect( sub { die "eval catch" if @_; } );

    eval { Package::foo(1); };
    like ($@, qr/eval catch/, "side_effect worked with params");

    # extract the parameters the sub was called with (if return_value or
    # side_effect is not used, we will return the parameters that were sent into
    # the mocked sub (list or scalar context)

    my @args = $foo->called_with;

    # reset the mock object for re-use within the same scope

    $foo->reset;

    # restore original functionality to the sub

    $foo->unmock;

    # re-mock a previously unmock()ed sub

    $foo->remock;

    # check if a sub is mocked

    my $state = $foo->mocked_state;

    # mock out a CORE:: function. Be warned that this *must* be done within
    # compile stage (BEGIN), and the function can NOT be unmocked prior
    # to the completion of program execution

    my ($mock, $caller);

    BEGIN {
        $mock = Mock::Sub->new;
        $caller = $mock->mock('caller');
    };

    $caller->return_value(55);
    caller(); # mocked caller() called

=head1 DESCRIPTION

Easy to use and very lightweight module for mocking out sub calls.
Very useful for testing areas of your own modules where getting coverage may
be difficult due to nothing to test against, and/or to reduce test run time by
eliminating the need to call subs that you really don't want or need to test.

=head1 EXAMPLE

Here's a full example to get further coverage where it's difficult if not
impossible to test certain areas of your code (eg: you have if/else statements,
but they don't do anything but call other subs. You don't want to test the
subs that are called, nor do you want to add statements to your code).

Note that if the end subroutine you're testing is NOT Object Oriented (and
you're importing them into your module that you're testing), you have to mock
them as part of your own namespace (ie. instead of Other::first, you'd mock
MyModule::first).

   # module you're testing:

    package MyPackage;
    
    use Other;
    use Exporter qw(import);
    @EXPORT_OK = qw(test);
   
    my $other = Other->new;

    sub test {
        my $arg = shift;
        
        if ($arg == 1){
            # how do you test this?... there's no return etc.
            $other->first();        
        }
        if ($arg == 2){
            $other->second();
        }
    }

    # your test file

    use MyPackage qw(test);
    use Mock::Sub;
    use Test::More tests => 2;

    my $mock = Mock::Sub->new;

    my $first = $mock->mock('Other::first');
    my $second = $mock->mock('Other::second');

    # coverage for first if() in MyPackage::test
    test(1);
    is ($first->called, 1, "1st if() statement covered");

    # coverage for second if()
    test(2);
    is ($second->called, 1, "2nd if() statement covered");

=head1 MOCK OBJECT METHODS

=head2 C<new(%opts)>

Instantiates and returns a new C<Mock::Sub> object, ready to be used to start
creating mocked sub objects.

Optional options:

=over 4

=item C<return_value =E<gt> $scalar>

Set this to have all mocked subs created with this mock object return anything
you wish (accepts a single scalar only. See C<return_value()> method to return
a list and for further information). You can also set it in individual mocks
only (see C<return_value()> method).

=item C<side_effect =E<gt> $cref>

Set this in C<new()> to have the side effect passed into all child mocks
created with this object. See C<side_effect()> method.

=back

=head2 C<mock('sub', %opts)>

Instantiates and returns a new mock object on each call. 'sub' is the name of
the subroutine to mock (requires full package name if the sub isn't in
C<main::>).

The mocked sub will return the parameters sent into the mocked sub if a return
value isn't set, or a side effect doesn't return anything, if available. If
in scalar context but a list was sent in, we'll return the first parameter in
the list. In list context, we simply receive the parameters as they were sent
in.

Optional parameters:

See C<new()> for a description of the parameters. Both the C<return_value> and
C<side_effect> parameters can be set in this method to individualize each mock
object, and will override the global configuration if set in C<new()>.

There's also C<return_value()> and C<side_effect()> methods if you want to
set, change or remove these values after instantiation of a child sub object.

=head2 mocked_subs

Returns a list of all the names of the subs that are currently mocked under
the parent mock object.

=head2 mocked_objects

Returns a list of all sub objects underneath the parent mock object, regardless
if its sub is currently mocked or not.

=head2 mocked_state('Sub::Name')

Returns 1 if the sub currently under the parent mock object is mocked or not,
and 0 if not. Croaks if there hasn't been a child sub object created with this
sub name.

=head1 SUB OBJECT METHODS

These methods are for the children mocked sub objects returned from the
parent mock object. See L<MOCK OBJECT METHODS> for methods related
to the parent mock object.

=head2 C<unmock>

Restores the original functionality back to the sub, and runs C<reset()> on
the object.

=head2 C<remock>

Re-mocks the sub within the object after calling C<unmock> on it (accepts the
side_effect and return_value parameters).

=head2 C<called>

Returns true (1) if the sub being mocked has been called, and false (0) if not.

=head2 C<called_count>

Returns the number of times the mocked sub has been called.

=head2 C<called_with>

Returns an array of the parameters sent to the subroutine. C<confess()s> if
we're called before the mocked sub has been called.

=head2 C<mocked_state>

Returns true (1) if the sub the object refers to is currently mocked, and
false (0) if not.

=head2 C<name>

Returns the name of the sub being mocked.

=head2 C<side_effect($cref)>

Add (or change/delete) a side effect after instantiation.

Send in a code reference containing an action you'd like the
mocked sub to perform.

The side effect function will receive all parameters sent into the mocked sub.

You can use both C<side_effect()> and C<return_value()> params at the same
time. C<side_effect> will be run first, and then C<return_value>. Note that if
C<side_effect>'s last expression evaluates to any value whatsoever
(even false), it will return that and C<return_value> will be skipped.

To work around this and have the side_effect run but still get the
return_value thereafter, write your cref to evaluate undef as the last thing
it does: C<sub { ...; undef; }>.

=head2 C<return_value>

Add (or change/delete) the mocked sub's return value after instantiation.
Can be a scalar or list. Send in C<undef> to remove previously set values.

=head2 C<reset>

Resets the functional parameters (C<return_value>, C<side_effect>), along
with C<called()> and C<called_count()> back to undef/false. Does not restore
the sub back to its original state.

=head1 NOTES

This module has a backwards parent-child relationship. To use, you create a
mock object using L</MOCK OBJECT METHODS> C<new> and C<mock> methods,
thereafter, you use the returned mocked sub object L</SUB OBJECT METHODS> to
perform the work.

The parent mock object retains certain information and statistics of the child
mocked objects (and the subs themselves).

To mock CORE::GLOBAL functions, you *must* initiate within a C<BEGIN> block
(see C<SYNOPSIS> for details). It is important that if you mock a CORE sub,
it can't and won't be returned to its original state until after the entire
program process tree exists. Period.

I didn't make this a C<Test::> module (although it started that way) because
I can see more uses than placing it into that category.

=head1 AUTHOR

Steve Bertrand, C<< <steveb at cpan.org> >>

=head1 BUGS

Please report any bugs or requests at
L<https://github.com/stevieb9/mock-sub/issues>

=head1 REPOSITORY

L<https://github.com/stevieb9/mock-sub>

=head1 BUILD RESULTS

CPAN Testers: L<http://matrix.cpantesters.org/?dist=Mock-Sub>

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Mock::Sub

=head1 ACKNOWLEDGEMENTS

Python's MagicMock module.

=head1 LICENSE AND COPYRIGHT

Copyright 2016 Steve Bertrand.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.


=cut

1; # End of Mock::Sub

